tests/stats_test.cpp

```diff
@@ -0,0 +1,390 @@
+}
+
+static void test_event_stats_with_time_dependency()
+{
+	const size_t default_size = 100;
+	const time_t default_time = time(nullptr);
+	stats_t stats(EVENTS_SIZE, PERIOD_IN_SECONDS);
+	stats_t stats2(EVENTS_SIZE, PERIOD_IN_SECONDS);
+	stats_t stats3(EVENTS_SIZE, PERIOD_IN_SECONDS);
+	std::vector<test_event> result, top_events;
+
+	// old events (outside observable period of time) are omitted
+	const size_t long_period = 10 * PERIOD_IN_SECONDS;
+	for(size_t i = 0; i < long_period; ++i) {
+		test_event e{"same", default_size, 1., default_time + static_cast<time_t>(i)};
+		stats.add_event(e, e.get_time());
+	}
+
+	stats.get_top(TOP_LENGTH, default_time + long_period, result);
+	BOOST_REQUIRE_EQUAL(result.size(), 1);
+	BOOST_CHECK(result.back().get_weight() <= PERIOD_IN_SECONDS * default_size);
+
+	// weight of event must be smaller after significant period of silence
+	test_event e{"same", default_size, 1., default_time + static_cast<time_t>(long_period + PERIOD_IN_SECONDS / 2)};
+	stats.add_event(e, e.get_time());
+	result.clear();	
+	stats.get_top(TOP_LENGTH, e.get_time(), result);
+	BOOST_REQUIRE_EQUAL(result.size(), 1);
+	BOOST_CHECK(result.back().get_weight() <= PERIOD_IN_SECONDS * default_size / 2);
+
+	// small key with regular frequent access must popup in top stats among heavier keys with single access
+	for(int i = 0; i < 3 * TOP_LENGTH; ++i) {
+		size_t size = i % 2 ? default_size : default_size * 10;
+		std::string key = i % 2 ? std::string("sum") : std::to_string(static_cast<long long>(i));
+
+		test_event e{key, size, 1., default_time + i};
+		stats2.add_event(e, e.get_time());
+	}
+
+	result.clear();	
+	stats2.get_top(TOP_LENGTH, default_time + 3 * TOP_LENGTH, result);
+	BOOST_REQUIRE_EQUAL(result.size(), TOP_LENGTH);
+	BOOST_REQUIRE_EQUAL(*result.front().get_key(), "sum");
+	BOOST_CHECK(result.front().get_weight() > result.back().get_weight());
+	BOOST_CHECK(result.front().get_frequency() > result.back().get_frequency());
+
+	// if keys have same size and time access pattern, then keys with more frequent access must be in top
+	for(int i = 0; i < 5 * TOP_LENGTH; ++i) {
+		std::string key = std::to_string(static_cast<long long>(i));
+		for(int j = 0; j < i+1; ++j) {
+			test_event e{key, default_size, 1., default_time + j};
+			stats3.add_event(e, e.get_time());
+		}
+		if (i >= 4 * TOP_LENGTH) {
+			test_event e{key, default_size, 1., default_time};
+			top_events.push_back(e);
+		}
+	}
+
+	result.clear();
+	stats3.get_top(TOP_LENGTH, default_time + 5 * TOP_LENGTH, result);
+	BOOST_REQUIRE_EQUAL(result.size(), TOP_LENGTH);
+	BOOST_REQUIRE_EQUAL(events_symmetric_diff(top_events, result), 0);
+}
+
+bool register_tests(test_suite *suite, node n)
+{
+	(void) n;
+	ELLIPTICS_TEST_CASE_NOARGS(test_top_provider_existance);
+	ELLIPTICS_TEST_CASE_NOARGS(test_event_stats_boundary_conditions);
+	ELLIPTICS_TEST_CASE_NOARGS(test_event_stats_no_time_dependency);
+	ELLIPTICS_TEST_CASE_NOARGS(test_event_stats_with_time_dependency);
```

Comment:
What is 'test_event' class for? Why does it copy-paste content of the cache processing?
What do tests do in details? Please describe it with more information about design of the tests, results, what is *not* supposed to be found in tests and what has to be there and so on 

From the first glance it looks like it compares several small (short-living) event sets fed into two event processors which are the same.

Do you have test data where you fed gigabytes of data, thousands of keys into the system and check for real that returned statistics matches the most actively downloaded keys?